File: roboco/api/routes/journals.py
"""
Journal API Routes

Agent personal journals for reflection, growth tracking, and debugging.

"""

from typing import Annotated
from uuid import UUID

from fastapi import APIRouter, Depends, HTTPException, status

from roboco.api.deps import CurrentAgentContext, DbSession
from roboco.api.schemas.journals import (
    CreateEntryRequest,
    DecisionLogRequest,
    GeneralEntryRequest,
    GrowthMetricsResponse,
    JournalEntryResponse,
    JournalResponse,
    JournalStatsResponse,
    LearningRequest,
    ListEntriesParams,
    SearchEntriesRequest,
    StruggleRequest,
    TaskReflectionRequest,
)
from roboco.enforcement import JournalAccessDeniedError, validate_journal_access
from roboco.models.base import JournalEntryType
from roboco.models.journal import (
    DecisionLogParams,
    GeneralEntryParams,
    JournalEntryCreate,
    LearningEntryParams,
    StruggleEntryParams,
    TaskReflectionParams,
)
from roboco.services.journal import ListEntriesFilter, get_journal_service

router = APIRouter()


# =============================================================================
# /me ENDPOINTS - MUST BE DEFINED BEFORE /{agent_id} ENDPOINTS
# =============================================================================


@router.get("/me", response_model=JournalResponse)
async def get_my_journal(
    agent: CurrentAgentContext,
    db: DbSession,
) -> JournalResponse:
    """Get or create the current agent's journal."""
    service = get_journal_service(db)
    journal = await service.get_or_create_journal(agent.agent_id)

    return JournalResponse(
        id=journal.id,
        agent_id=journal.agent_id,
        total_entries=journal.total_entries,
        last_entry_at=journal.last_entry_at,
        latest_summary=journal.latest_summary,
        summary_updated_at=journal.summary_updated_at,
        entries_by_type=journal.entries_by_type,
        created_at=journal.created_at,
        updated_at=journal.updated_at,
    )


@router.get("/me/entries", response_model=list[JournalEntryResponse])
async def list_my_entries(
    agent: CurrentAgentContext,
    db: DbSession,
    params: Annotated[ListEntriesParams, Depends()],
) -> list[JournalEntryResponse]:
    """List the current agent's journal entries."""
    service = get_journal_service(db)
    journal = await service.get_journal_by_agent(agent.agent_id)

    if not journal:
        return []

    type_filter = None
    if params.entry_type:
        try:
            type_filter = JournalEntryType(params.entry_type)
        except ValueError as e:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail=f"Invalid entry type: {e}",
            ) from e

    entries = await service.list_entries(
        journal_id=journal.id,
        filters=ListEntriesFilter(
            entry_type=type_filter,
            task_id=params.task_id,
            limit=params.limit,
            offset=params.offset,
            include_private=True,  # Can see own private entries
        ),
    )

    return [
        JournalEntryResponse(
            id=e.id,
            journal_id=e.journal_id,
            type=e.type.value if isinstance(e.type, JournalEntryType) else e.type,
            title=e.title,
            content=e.content,
            task_id=e.task_id,
            session_id=e.session_id,
            timestamp=e.timestamp,
            tags=e.tags,
            sentiment=e.sentiment,
            is_private=e.is_private,
            created_at=e.created_at,
            updated_at=e.updated_at,
        )
        for e in entries
    ]


@router.post(
    "/me/entries",
    response_model=JournalEntryResponse,
    status_code=status.HTTP_201_CREATED,
)
async def create_entry(
    request: CreateEntryRequest,
    agent: CurrentAgentContext,
    db: DbSession,
) -> JournalEntryResponse:
    """Create a new journal entry."""
    service = get_journal_service(db)
    journal = await service.get_or_create_journal(agent.agent_id)

    try:
        entry_type = JournalEntryType(request.type)
    except ValueError as e:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=f"Invalid entry type: {e}",
        ) from e

    entry_create = JournalEntryCreate(
        journal_id=journal.id,
        type=entry_type,
        title=request.title,
        content=request.content,
        task_id=request.task_id,
        session_id=request.session_id,
        tags=request.tags,
        sentiment=request.sentiment,
        is_private=request.is_private,
    )

    entry = await service.create_entry(entry_create)

    return JournalEntryResponse(
        id=entry.id,
        journal_id=entry.journal_id,
        type=entry.type.value
        if isinstance(entry.type, JournalEntryType)
        else entry.type,
        title=entry.title,
        content=entry.content,
        task_id=entry.task_id,
        session_id=entry.session_id,
        timestamp=entry.timestamp,
        tags=entry.tags,
        sentiment=entry.sentiment,
        is_private=entry.is_private,
        created_at=entry.created_at,
        updated_at=entry.updated_at,
    )


@router.get("/me/stats", response_model=JournalStatsResponse)
async def get_my_stats(
    agent: CurrentAgentContext,
    db: DbSession,
) -> JournalStatsResponse:
    """Get statistics for the current agent's journal."""
    service = get_journal_service(db)
    journal = await service.get_journal_by_agent(agent.agent_id)

    if not journal:
        return JournalStatsResponse(
            total_entries=0,
            entries_by_type={},
            last_entry_at=None,
            has_summary=False,
        )

    stats = await service.get_journal_stats(journal.id)
    if not stats:
        return JournalStatsResponse(
            total_entries=0,
            entries_by_type={},
            last_entry_at=None,
            has_summary=False,
        )

    return JournalStatsResponse(
        total_entries=stats.total_entries,
        entries_by_type=stats.entries_by_type,
        last_entry_at=stats.last_entry_at,
        has_summary=stats.has_summary,
    )


@router.get("/me/growth", response_model=GrowthMetricsResponse)
async def get_my_growth_metrics(
    agent: CurrentAgentContext,
    db: DbSession,
) -> GrowthMetricsResponse:
    """Get growth metrics for the current agent."""
    service = get_journal_service(db)
    metrics = await service.get_growth_metrics(agent.agent_id)

    if not metrics:
        return GrowthMetricsResponse(
            total_reflections=0,
            total_learnings=0,
            total_struggles=0,
            total_decisions=0,
            struggle_resolution_rate=0.0,
            learning_frequency=0.0,
            sentiment_trend="stable",
        )

    return GrowthMetricsResponse(
        total_reflections=metrics.total_reflections,
        total_learnings=metrics.total_learnings,
        total_struggles=metrics.total_struggles,
        total_decisions=metrics.total_decisions,
        struggle_resolution_rate=metrics.struggle_resolution_rate,
        learning_frequency=metrics.learning_frequency,
        sentiment_trend=metrics.sentiment_trend,
    )


@router.post("/me/search", response_model=list[JournalEntryResponse])
async def search_my_entries(
    request: SearchEntriesRequest,
    agent: CurrentAgentContext,
    db: DbSession,
) -> list[JournalEntryResponse]:
    """
    Semantic search through the current agent's journal.

    Uses RAG to find relevant entries.
    """
    service = get_journal_service(db)
    entries = await service.search_entries(
        agent_id=agent.agent_id,
        query=request.query,
        top_k=request.top_k,
    )

    return [
        JournalEntryResponse(
            id=e.id,
            journal_id=e.journal_id,
            type=e.type.value if isinstance(e.type, JournalEntryType) else e.type,
            title=e.title,
            content=e.content,
            task_id=e.task_id,
            session_id=e.session_id,
            timestamp=e.timestamp,
            tags=e.tags,
            sentiment=e.sentiment,
            is_private=e.is_private,
            created_at=e.created_at,
            updated_at=e.updated_at,
        )
        for e in entries
    ]


@router.post(
    "/me/reflections",
    response_model=JournalEntryResponse,
    status_code=status.HTTP_201_CREATED,
)
async def add_task_reflection(
    request: TaskReflectionRequest,
    agent: CurrentAgentContext,
    db: DbSession,
) -> JournalEntryResponse:
    """Add a task reflection entry."""
    service = get_journal_service(db)
    entry = await service.add_task_reflection(
        agent_id=agent.agent_id,
        params=TaskReflectionParams(
            task_id=request.task_id,
            title=request.title,
            what_done=request.what_done,
            what_learned=request.what_learned,
            what_struggled=request.what_struggled,
            next_steps=request.next_steps,
            tags=request.tags,
        ),
    )

    return JournalEntryResponse(
        id=entry.id,
        journal_id=entry.journal_id,
        type=entry.type.value
        if isinstance(entry.type, JournalEntryType)
        else entry.type,
        title=entry.title,
        content=entry.content,
        task_id=entry.task_id,
        session_id=entry.session_id,
        timestamp=entry.timestamp,
        tags=entry.tags,
        sentiment=entry.sentiment,
        is_private=entry.is_private,
        created_at=entry.created_at,
    )


@router.post(
    "/me/decisions",
    response_model=JournalEntryResponse,
    status_code=status.HTTP_201_CREATED,
)
async def add_decision_log(
    request: DecisionLogRequest,
    agent: CurrentAgentContext,
    db: DbSession,
) -> JournalEntryResponse:
    """Add a decision log entry."""
    service = get_journal_service(db)
    entry = await service.add_decision_log(
        agent_id=agent.agent_id,
        params=DecisionLogParams(
            title=request.title,
            context=request.context,
            options=request.options,
            chosen=request.chosen,
            rationale=request.rationale,
            consequences=request.consequences,
            task_id=request.task_id,
            tags=request.tags,
        ),
    )

    return JournalEntryResponse(
        id=entry.id,
        journal_id=entry.journal_id,
        type=entry.type.value
        if isinstance(entry.type, JournalEntryType)
        else entry.type,
        title=entry.title,
        content=entry.content,
        task_id=entry.task_id,
        session_id=entry.session_id,
        timestamp=entry.timestamp,
        tags=entry.tags,
        sentiment=entry.sentiment,
        is_private=entry.is_private,
        created_at=entry.created_at,
    )


@router.post(
    "/me/learnings",
    response_model=JournalEntryResponse,
    status_code=status.HTTP_201_CREATED,
)
async def add_learning(
    request: LearningRequest,
    agent: CurrentAgentContext,
    db: DbSession,
) -> JournalEntryResponse:
    """Add a learning entry."""
    service = get_journal_service(db)
    entry = await service.add_learning(
        agent_id=agent.agent_id,
        params=LearningEntryParams(
            title=request.title,
            what_learned=request.what_learned,
            how_applied=request.how_applied,
            source=request.source,
            task_id=request.task_id,
            tags=request.tags,
        ),
    )

    return JournalEntryResponse(
        id=entry.id,
        journal_id=entry.journal_id,
        type=entry.type.value
        if isinstance(entry.type, JournalEntryType)
        else entry.type,
        title=entry.title,
        content=entry.content,
        task_id=entry.task_id,
        session_id=entry.session_id,
        timestamp=entry.timestamp,
        tags=entry.tags,
        sentiment=entry.sentiment,
        is_private=entry.is_private,
        created_at=entry.created_at,
    )


@router.post(
    "/me/struggles",
    response_model=JournalEntryResponse,
    status_code=status.HTTP_201_CREATED,
)
async def add_struggle(
    request: StruggleRequest,
    agent: CurrentAgentContext,
    db: DbSession,
) -> JournalEntryResponse:
    """Add a struggle entry."""
    service = get_journal_service(db)
    entry = await service.add_struggle(
        agent_id=agent.agent_id,
        params=StruggleEntryParams(
            title=request.title,
            what_struggled=request.what_struggled,
            attempted_solutions=request.attempted_solutions,
            resolution=request.resolution,
            help_needed=request.help_needed,
            task_id=request.task_id,
            tags=request.tags,
        ),
    )

    return JournalEntryResponse(
        id=entry.id,
        journal_id=entry.journal_id,
        type=entry.type.value
        if isinstance(entry.type, JournalEntryType)
        else entry.type,
        title=entry.title,
        content=entry.content,
        task_id=entry.task_id,
        session_id=entry.session_id,
        timestamp=entry.timestamp,
        tags=entry.tags,
        sentiment=entry.sentiment,
        is_private=entry.is_private,
        created_at=entry.created_at,
    )


@router.post(
    "/me/notes",
    response_model=JournalEntryResponse,
    status_code=status.HTTP_201_CREATED,
)
async def add_general_entry(
    request: GeneralEntryRequest,
    agent: CurrentAgentContext,
    db: DbSession,
) -> JournalEntryResponse:
    """Add a general journal entry."""
    service = get_journal_service(db)
    entry = await service.add_general_entry(
        agent_id=agent.agent_id,
        params=GeneralEntryParams(
            title=request.title,
            content=request.content,
            task_id=request.task_id,
            session_id=request.session_id,
            tags=request.tags,
            is_private=request.is_private,
        ),
    )

    return JournalEntryResponse(
        id=entry.id,
        journal_id=entry.journal_id,
        type=entry.type.value
        if isinstance(entry.type, JournalEntryType)
        else entry.type,
        title=entry.title,
        content=entry.content,
        task_id=entry.task_id,
        session_id=entry.session_id,
        timestamp=entry.timestamp,
        tags=entry.tags,
        sentiment=entry.sentiment,
        is_private=entry.is_private,
        created_at=entry.created_at,
    )


# =============================================================================
# ENTRY CRUD ENDPOINTS (non-/me paths)
# =============================================================================


@router.get("/entries/{entry_id}", response_model=JournalEntryResponse)
async def get_entry(
    entry_id: UUID,
    agent: CurrentAgentContext,
    db: DbSession,
) -> JournalEntryResponse:
    """Get a specific journal entry."""
    service = get_journal_service(db)
    entry = await service.get_entry(entry_id)

    if not entry:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Entry not found: {entry_id}",
        )

    # Get journal to check ownership
    journal = await service.get_journal(entry.journal_id)
    if not journal:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Journal not found for entry: {entry_id}",
        )

    # Get owner's slug for permission checking
    owner_slug = await service.get_agent_slug(journal.agent_id)
    if not owner_slug:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Journal owner not found",
        )

    # Check permission (cell members can see all entries including private)
    try:
        validate_journal_access(agent.slug or "", owner_slug)
    except JournalAccessDeniedError as e:
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail=e.message,
        ) from e

    return JournalEntryResponse(
        id=entry.id,
        journal_id=entry.journal_id,
        type=entry.type.value
        if isinstance(entry.type, JournalEntryType)
        else entry.type,
        title=entry.title,
        content=entry.content,
        task_id=entry.task_id,
        session_id=entry.session_id,
        timestamp=entry.timestamp,
        tags=entry.tags,
        sentiment=entry.sentiment,
        is_private=entry.is_private,
        created_at=entry.created_at,
        updated_at=entry.updated_at,
    )


@router.delete("/entries/{entry_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_entry(
    entry_id: UUID,
    agent: CurrentAgentContext,
    db: DbSession,
) -> None:
    """Delete a journal entry."""
    service = get_journal_service(db)
    entry = await service.get_entry(entry_id)

    if not entry:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Entry not found: {entry_id}",
        )

    # Check ownership
    journal = await service.get_journal(entry.journal_id)
    if journal and journal.agent_id != agent.agent_id:
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="Can only delete your own entries",
        )

    await service.delete_entry(entry_id)


# =============================================================================
# /{agent_id} ENDPOINTS - MUST BE DEFINED AFTER /me ENDPOINTS
# =============================================================================


@router.get("/{agent_id}", response_model=JournalResponse)
async def get_journal_by_agent(
    agent_id: str,
    agent: CurrentAgentContext,
    db: DbSession,
) -> JournalResponse:
    """
    Get a journal by agent ID (UUID) or slug (e.g., "be-dev-1").

    Access is restricted based on cell membership and role hierarchy.
    """
    service = get_journal_service(db)

    # Resolve agent identifier (UUID or slug) to UUID
    resolved_id = await service.resolve_agent_id(agent_id)
    if not resolved_id:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Agent not found: {agent_id}",
        )

    # Get target agent's slug for permission checking
    target_slug = await service.get_agent_slug(resolved_id)
    if not target_slug:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Agent not found: {agent_id}",
        )

    # Check permission
    try:
        validate_journal_access(agent.slug or "", target_slug)
    except JournalAccessDeniedError as e:
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail=e.message,
        ) from e

    journal = await service.get_journal_by_agent(resolved_id)
    if not journal:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Journal not found for agent: {agent_id}",
        )

    return JournalResponse(
        id=journal.id,
        agent_id=journal.agent_id,
        total_entries=journal.total_entries,
        last_entry_at=journal.last_entry_at,
        latest_summary=journal.latest_summary,
        summary_updated_at=journal.summary_updated_at,
        entries_by_type=journal.entries_by_type,
        created_at=journal.created_at,
        updated_at=journal.updated_at,
    )


@router.get("/{agent_id}/entries", response_model=list[JournalEntryResponse])
async def list_agent_entries(
    agent_id: str,
    agent: CurrentAgentContext,
    db: DbSession,
    params: Annotated[ListEntriesParams, Depends()],
) -> list[JournalEntryResponse]:
    """
    List journal entries for a specific agent by UUID or slug.

    Access is restricted based on cell membership and role hierarchy.
    Cell members can see all entries (including private) from each other.
    """
    service = get_journal_service(db)

    # Resolve agent identifier (UUID or slug) to UUID
    resolved_id = await service.resolve_agent_id(agent_id)
    if not resolved_id:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Agent not found: {agent_id}",
        )

    # Get target agent's slug for permission checking
    target_slug = await service.get_agent_slug(resolved_id)
    if not target_slug:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Agent not found: {agent_id}",
        )

    # Check permission
    try:
        validate_journal_access(agent.slug or "", target_slug)
    except JournalAccessDeniedError as e:
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail=e.message,
        ) from e

    journal = await service.get_journal_by_agent(resolved_id)
    if not journal:
        return []

    type_filter = None
    if params.entry_type:
        try:
            type_filter = JournalEntryType(params.entry_type)
        except ValueError as e:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail=f"Invalid entry type: {e}",
            ) from e

    # Cell members with access can see all entries including private
    entries = await service.list_entries(
        journal_id=journal.id,
        filters=ListEntriesFilter(
            entry_type=type_filter,
            task_id=params.task_id,
            limit=params.limit,
            offset=params.offset,
            include_private=True,  # Full access for authorized readers
        ),
    )

    return [
        JournalEntryResponse(
            id=e.id,
            journal_id=e.journal_id,
            type=e.type.value if isinstance(e.type, JournalEntryType) else e.type,
            title=e.title,
            content=e.content,
            task_id=e.task_id,
            session_id=e.session_id,
            timestamp=e.timestamp,
            tags=e.tags,
            sentiment=e.sentiment,
            is_private=e.is_private,
            created_at=e.created_at,
            updated_at=e.updated_at,
        )
        for e in entries
    ]

File: panel/src/lib/api/journals.ts
/**
 * Journals API Client
 *
 * Agent personal journals for reflection, growth tracking, and debugging.
 * Matches backend: roboco/api/routes/journals.py
 */

import api from "./client";
import {
  Journal,
  JournalEntry,
  JournalEntryCreate,
  JournalEntryType,
  JournalStats,
  GrowthMetrics,
} from "@/types";
import {
  isMockMode,
  mockJournals,
  mockJournalEntries,
} from "@/lib/mock-data";

// =============================================================================
// JOURNAL ENDPOINTS
// =============================================================================

/**
 * Get the current agent's journal (CEO's journal in this case)
 */
async function getMyJournal(): Promise<Journal> {
  if (isMockMode()) {
    return mockJournals[0] as Journal;
  }
  const response = await api.get<Journal>("/journals/me");
  return response.data;
}

/**
 * Get a journal by agent ID or slug (e.g., "be-dev-1")
 */
async function getJournalByAgent(agentIdOrSlug: string): Promise<Journal> {
  if (isMockMode()) {
    const journal = mockJournals.find((j) => j.agent_id === agentIdOrSlug);
    if (journal) return journal as Journal;
    // Return first journal as fallback
    return mockJournals[0] as Journal;
  }
  const response = await api.get<Journal>(`/journals/${agentIdOrSlug}`);
  return response.data;
}

/**
 * List entries for a specific agent by ID or slug
 */
async function listAgentEntries(
  agentIdOrSlug: string,
  params?: {
    entry_type?: JournalEntryType;
    task_id?: string;
    limit?: number;
    offset?: number;
  }
): Promise<JournalEntry[]> {
  if (isMockMode()) {
    let entries = [...mockJournalEntries] as JournalEntry[];
    if (params?.entry_type) {
      entries = entries.filter((e) => e.type === params.entry_type);
    }
    if (params?.task_id) {
      entries = entries.filter((e) => e.task_id === params.task_id);
    }
    const offset = params?.offset ?? 0;
    const limit = params?.limit ?? 50;
    return entries.slice(offset, offset + limit);
  }
  const response = await api.get<JournalEntry[]>(
    `/journals/${agentIdOrSlug}/entries`,
    { params }
  );
  return response.data;
}

// =============================================================================
// ENTRY CRUD ENDPOINTS
// =============================================================================

/**
 * Create a new journal entry
 */
async function createEntry(data: JournalEntryCreate): Promise<JournalEntry> {
  if (isMockMode()) {
    const newEntry: JournalEntry = {
      id: `entry-${Date.now()}`,
      journal_id: mockJournals[0].id,
      type: data.type,
      title: data.title,
      content: data.content,
      task_id: data.task_id ?? null,
      session_id: data.session_id ?? null,
      timestamp: new Date().toISOString(),
      tags: data.tags ?? [],
      sentiment: data.sentiment ?? null,
      is_private: data.is_private ?? false,
      created_at: new Date().toISOString(),
      updated_at: null,
    };
    (mockJournalEntries as JournalEntry[]).push(newEntry);
    return newEntry;
  }
  const response = await api.post<JournalEntry>("/journals/me/entries", data);
  return response.data;
}

/**
 * List current agent's journal entries
 */
async function listMyEntries(params?: {
  entry_type?: JournalEntryType;
  task_id?: string;
  limit?: number;
  offset?: number;
}): Promise<JournalEntry[]> {
  if (isMockMode()) {
    let entries = [...mockJournalEntries] as JournalEntry[];
    if (params?.entry_type) {
      entries = entries.filter((e) => e.type === params.entry_type);
    }
    if (params?.task_id) {
      entries = entries.filter((e) => e.task_id === params.task_id);
    }
    const offset = params?.offset ?? 0;
    const limit = params?.limit ?? 50;
    return entries.slice(offset, offset + limit);
  }
  const response = await api.get<JournalEntry[]>("/journals/me/entries", {
    params,
  });
  return response.data;
}

/**
 * Get a specific journal entry by ID
 */
async function getEntry(entryId: string): Promise<JournalEntry> {
  if (isMockMode()) {
    const entry = mockJournalEntries.find((e) => e.id === entryId);
    if (entry) return entry as JournalEntry;
    throw new Error("Entry not found");
  }
  const response = await api.get<JournalEntry>(`/journals/entries/${entryId}`);
  return response.data;
}

/**
 * Delete a journal entry
 */
async function deleteEntry(entryId: string): Promise<void> {
  if (isMockMode()) {
    const idx = mockJournalEntries.findIndex((e) => e.id === entryId);
    if (idx !== -1) mockJournalEntries.splice(idx, 1);
    return;
  }
  await api.delete(`/journals/entries/${entryId}`);
}

// =============================================================================
// CONVENIENCE ENTRY ENDPOINTS
// =============================================================================

/**
 * Add a task reflection entry
 */
async function addTaskReflection(data: {
  task_id: string;
  title: string;
  what_done: string;
  what_learned: string;
  what_struggled?: string;
  next_steps?: string[];
  tags?: string[];
}): Promise<JournalEntry> {
  if (isMockMode()) {
    return createEntry({
      type: JournalEntryType.TASK_REFLECTION,
      title: data.title,
      content: JSON.stringify({
        what_done: data.what_done,
        what_learned: data.what_learned,
        what_struggled: data.what_struggled,
        next_steps: data.next_steps,
      }),
      task_id: data.task_id,
      tags: data.tags,
    });
  }
  const response = await api.post<JournalEntry>(
    "/journals/me/reflections",
    data
  );
  return response.data;
}

/**
 * Add a decision log entry
 */
async function addDecisionLog(data: {
  title: string;
  context: string;
  options: string[];
  chosen: string;
  rationale: string;
  consequences?: string;
  task_id?: string;
  tags?: string[];
}): Promise<JournalEntry> {
  if (isMockMode()) {
    return createEntry({
      type: JournalEntryType.DECISION_LOG,
      title: data.title,
      content: JSON.stringify({
        context: data.context,
        options: data.options,
        chosen: data.chosen,
        rationale: data.rationale,
        consequences: data.consequences,
      }),
      task_id: data.task_id,
      tags: data.tags,
    });
  }
  const response = await api.post<JournalEntry>("/journals/me/decisions", data);
  return response.data;
}

/**
 * Add a learning entry
 */
async function addLearning(data: {
  title: string;
  what_learned: string;
  how_applied?: string;
  source?: string;
  task_id?: string;
  tags?: string[];
}): Promise<JournalEntry> {
  if (isMockMode()) {
    return createEntry({
      type: JournalEntryType.LEARNING,
      title: data.title,
      content: JSON.stringify({
        what_learned: data.what_learned,
        how_applied: data.how_applied,
        source: data.source,
      }),
      task_id: data.task_id,
      tags: data.tags,
    });
  }
  const response = await api.post<JournalEntry>("/journals/me/learnings", data);
  return response.data;
}

/**
 * Add a struggle entry
 */
async function addStruggle(data: {
  title: string;
  what_struggled: string;
  attempted_solutions: string[];
  resolution?: string;
  help_needed?: string;
  task_id?: string;
  tags?: string[];
}): Promise<JournalEntry> {
  if (isMockMode()) {
    return createEntry({
      type: JournalEntryType.STRUGGLE,
      title: data.title,
      content: JSON.stringify({
        what_struggled: data.what_struggled,
        attempted_solutions: data.attempted_solutions,
        resolution: data.resolution,
        help_needed: data.help_needed,
      }),
      task_id: data.task_id,
      tags: data.tags,
    });
  }
  const response = await api.post<JournalEntry>("/journals/me/struggles", data);
  return response.data;
}

/**
 * Add a general journal entry
 */
async function addNote(data: {
  title: string;
  content: string;
  task_id?: string;
  session_id?: string;
  tags?: string[];
  is_private?: boolean;
}): Promise<JournalEntry> {
  if (isMockMode()) {
    return createEntry({
      type: JournalEntryType.GENERAL,
      title: data.title,
      content: data.content,
      task_id: data.task_id,
      session_id: data.session_id,
      tags: data.tags,
      is_private: data.is_private,
    });
  }
  const response = await api.post<JournalEntry>("/journals/me/notes", data);
  return response.data;
}

// =============================================================================
// ANALYTICS ENDPOINTS
// =============================================================================

/**
 * Get statistics for the current agent's journal
 */
async function getMyStats(): Promise<JournalStats> {
  if (isMockMode()) {
    const entries = mockJournalEntries as JournalEntry[];
    return {
      total_entries: entries.length,
      entries_by_type: {
        task_reflection: entries.filter((e) => e.type === JournalEntryType.TASK_REFLECTION).length,
        decision_log: entries.filter((e) => e.type === JournalEntryType.DECISION_LOG).length,
        learning: entries.filter((e) => e.type === JournalEntryType.LEARNING).length,
        struggle: entries.filter((e) => e.type === JournalEntryType.STRUGGLE).length,
        general: entries.filter((e) => e.type === JournalEntryType.GENERAL).length,
      },
      last_entry_at: entries[0]?.created_at ?? null,
      has_summary: false,
    };
  }
  const response = await api.get<JournalStats>("/journals/me/stats");
  return response.data;
}

/**
 * Get growth metrics for the current agent
 */
async function getMyGrowthMetrics(): Promise<GrowthMetrics> {
  if (isMockMode()) {
    const entries = mockJournalEntries as JournalEntry[];
    return {
      total_reflections: entries.filter((e) => e.type === JournalEntryType.TASK_REFLECTION).length,
      total_learnings: entries.filter((e) => e.type === JournalEntryType.LEARNING).length,
      total_struggles: entries.filter((e) => e.type === JournalEntryType.STRUGGLE).length,
      total_decisions: entries.filter((e) => e.type === JournalEntryType.DECISION_LOG).length,
      struggle_resolution_rate: 0.7,
      learning_frequency: 3.5,
      sentiment_trend: "improving",
    };
  }
  const response = await api.get<GrowthMetrics>("/journals/me/growth");
  return response.data;
}

// =============================================================================
// SEARCH ENDPOINTS
// =============================================================================

/**
 * Semantic search through journal entries
 */
async function searchEntries(
  query: string,
  topK: number = 10
): Promise<JournalEntry[]> {
  if (isMockMode()) {
    // Simple text search for mock mode
    const queryLower = query.toLowerCase();
    return (mockJournalEntries as JournalEntry[])
      .filter(
        (e) =>
          e.title.toLowerCase().includes(queryLower) ||
          e.content.toLowerCase().includes(queryLower)
      )
      .slice(0, topK);
  }
  const response = await api.post<JournalEntry[]>("/journals/me/search", {
    query,
    top_k: topK,
  });
  return response.data;
}

// =============================================================================
// EXPORT API
// =============================================================================

export const journalsApi = {
  // Journal
  getMyJournal,
  getJournalByAgent,

  // Entry CRUD
  createEntry,
  listMyEntries,
  listAgentEntries,
  getEntry,
  deleteEntry,

  // Convenience entry creators
  addTaskReflection,
  addDecisionLog,
  addLearning,
  addStruggle,
  addNote,

  // Analytics
  getMyStats,
  getMyGrowthMetrics,

  // Search
  searchEntries,
};

export default journalsApi;
